### PR TITLE
Lint with `flake8-bugbear`

### PR DESCRIPTION
**Pull request by danieleades**
adds 'flake8-bugbear' linting.

This is throwing a large number of errors, related to using mutable argument defaults. Would be really handy to have a maintainer comment on whether the large number of mutable argument defaults is indeed deliberate, or if it truly is an oversight. From what I can see, in most of the examples below, there is a default list that gets iterated over in the function, but not modified, so it works. However it's not a great pattern, because if at any point the code was refactored in such a way that a list was no longer treated as read-only, then you would silently introduce some very strange errors. I've not reviewed each of these detail, so it's entirely possible that's already the case in some of them.

<details>
  <summary>mutable argument defaults errors:</summary>

```
./sphinx/theming.py:108:55: B006 Do not use mutable data structures for argument defaults.  They are created during function definition time. All calls to the function reuse this one instance of that data structure, persisting changes between them.
./sphinx/registry.py:233:68: B006 Do not use mutable data structures for argument defaults.  They are created during function definition time. All calls to the function reuse this one instance of that data structure, persisting changes between them.
./sphinx/config.py:146:49: B006 Do not use mutable data structures for argument defaults.  They are created during function definition time. All calls to the function reuse this one instance of that data structure, persisting changes between them.
./sphinx/config.py:146:81: B006 Do not use mutable data structures for argument defaults.  They are created during function definition time. All calls to the function reuse this one instance of that data structure, persisting changes between them.
./sphinx/application.py:794:68: B006 Do not use mutable data structures for argument defaults.  They are created during function definition time. All calls to the function reuse this one instance of that data structure, persisting changes between them.
./sphinx/project.py:33:51: B006 Do not use mutable data structures for argument defaults.  They are created during function definition time. All calls to the function reuse this one instance of that data structure, persisting changes between them.
./sphinx/deprecation.py:21:70: B006 Do not use mutable data structures for argument defaults.  They are created during function definition time. All calls to the function reuse this one instance of that data structure, persisting changes between them.
./sphinx/roles.py:365:31: B006 Do not use mutable data structures for argument defaults.  They are created during function definition time. All calls to the function reuse this one instance of that data structure, persisting changes between them.
./sphinx/roles.py:365:56: B006 Do not use mutable data structures for argument defaults.  They are created during function definition time. All calls to the function reuse this one instance of that data structure, persisting changes between them.
./sphinx/ext/extlinks.py:110:48: B006 Do not use mutable data structures for argument defaults.  They are created during function definition time. All calls to the function reuse this one instance of that data structure, persisting changes between them.
./sphinx/ext/extlinks.py:110:73: B006 Do not use mutable data structures for argument defaults.  They are created during function definition time. All calls to the function reuse this one instance of that data structure, persisting changes between them.
./sphinx/ext/apidoc.py:110:47: B006 Do not use mutable data structures for argument defaults.  They are created during function definition time. All calls to the function reuse this one instance of that data structure, persisting changes between them.
./sphinx/ext/apidoc.py:168:71: B006 Do not use mutable data structures for argument defaults.  They are created during function definition time. All calls to the function reuse this one instance of that data structure, persisting changes between them.
./sphinx/ext/inheritance_diagram.py:135:43: B006 Do not use mutable data structures for argument defaults.  They are created during function definition time. All calls to the function reuse this one instance of that data structure, persisting changes between them.
./sphinx/ext/inheritance_diagram.py:265:52: B006 Do not use mutable data structures for argument defaults.  They are created during function definition time. All calls to the function reuse this one instance of that data structure, persisting changes between them.
./sphinx/ext/inheritance_diagram.py:266:42: B006 Do not use mutable data structures for argument defaults.  They are created during function definition time. All calls to the function reuse this one instance of that data structure, persisting changes between them.
./sphinx/ext/inheritance_diagram.py:266:65: B006 Do not use mutable data structures for argument defaults.  They are created during function definition time. All calls to the function reuse this one instance of that data structure, persisting changes between them.
./sphinx/ext/inheritance_diagram.py:266:88: B006 Do not use mutable data structures for argument defaults.  They are created during function definition time. All calls to the function reuse this one instance of that data structure, persisting changes between them.
./sphinx/ext/autosummary/generate.py:248:76: B006 Do not use mutable data structures for argument defaults.  They are created during function definition time. All calls to the function reuse this one instance of that data structure, persisting changes between them.
./sphinx/ext/autosummary/__init__.py:625:53: B006 Do not use mutable data structures for argument defaults.  They are created during function definition time. All calls to the function reuse this one instance of that data structure, persisting changes between them.
./sphinx/ext/autosummary/__init__.py:706:58: B006 Do not use mutable data structures for argument defaults.  They are created during function definition time. All calls to the function reuse this one instance of that data structure, persisting changes between them.
./sphinx/ext/autodoc/type_comment.py:17:46: B006 Do not use mutable data structures for argument defaults.  They are created during function definition time. All calls to the function reuse this one instance of that data structure, persisting changes between them.
./sphinx/ext/napoleon/docstring.py:49:67: B006 Do not use mutable data structures for argument defaults.  They are created during function definition time. All calls to the function reuse this one instance of that data structure, persisting changes between them.
./sphinx/ext/napoleon/docstring.py:1002:85: B006 Do not use mutable data structures for argument defaults.  They are created during function definition time. All calls to the function reuse this one instance of that data structure, persisting changes between them.
./sphinx/writers/text.py:404:61: B006 Do not use mutable data structures for argument defaults.  They are created during function definition time. All calls to the function reuse this one instance of that data structure, persisting changes between them.
./sphinx/domains/__init__.py:260:60: B006 Do not use mutable data structures for argument defaults.  They are created during function definition time. All calls to the function reuse this one instance of that data structure, persisting changes between them.
./sphinx/domains/__init__.py:260:85: B006 Do not use mutable data structures for argument defaults.  They are created during function definition time. All calls to the function reuse this one instance of that data structure, persisting changes between them.
./sphinx/util/inspect.py:560:83: B006 Do not use mutable data structures for argument defaults.  They are created during function definition time. All calls to the function reuse this one instance of that data structure, persisting changes between them.
./sphinx/util/docutils.py:415:52: B006 Do not use mutable data structures for argument defaults.  They are created during function definition time. All calls to the function reuse this one instance of that data structure, persisting changes between them.
./sphinx/util/docutils.py:415:77: B006 Do not use mutable data structures for argument defaults.  They are created during function definition time. All calls to the function reuse this one instance of that data structure, persisting changes between them.
./sphinx/util/docutils.py:478:52: B006 Do not use mutable data structures for argument defaults.  They are created during function definition time. All calls to the function reuse this one instance of that data structure, persisting changes between them.
./sphinx/util/docutils.py:478:77: B006 Do not use mutable data structures for argument defaults.  They are created during function definition time. All calls to the function reuse this one instance of that data structure, persisting changes between them.
./tests/test_ext_autodoc.py:410:55: B006 Do not use mutable data structures for argument defaults.  They are created during function definition time. All calls to the function reuse this one instance of that data structure, persisting changes between them.
./tests/typing_test_data.py:42:33: B006 Do not use mutable data structures for argument defaults.  They are created during function definition time. All calls to the function reuse this one instance of that data structure, persisting changes between them.
```
</details>

### CODE REVIEW

**Comment by AA-Turner** [comment]
Tests are failing right now, but if you can get tests passing I'm happy to merge this.

Mutable: I agree it is an anti-pattern, and we can either use an empty tuple to signal a `Sequence` type is expected, or other refactoring. I'd prefer if that was a different PR though, please.

A

**Comment by danieleades** [comment]
> Tests are failing right now, but if you can get tests passing I'm happy to merge this.
> 
> Mutable: I agree it is an anti-pattern, and we can either use an empty tuple to signal a `Sequence` type is expected, or other refactoring. I'd prefer if that was a different PR though, please.
> 
> A

i've silenced the lint for now. to be addressed in another PR

**Comment by danieleades** [comment]
note that this PR introduces changed that mean that assertions will *not* be stripped when running python with optimisations enabled.
@AA-Turner I just want to confirm this behaviour is desirable.

**Comment by AA-Turner** [comment]
I'll check when I review, but in general I would say yes, using `assert cond` is undesirable except in test code. If there is a safety check, we should use `__debug__` rather than `assert cond`.

A

**Review by tk0miya — APPROVED** [review]
LGTM

**Review by AA-Turner — APPROVED** [review]
Changes look good, but there are several in domains/c(pp).py, so I'd like @jakobandersen to review before merging -- running with `python -O` is currently a way to get out of those asserts if you trigger them, especially the ones with "can this happen" etc.

A

**Comment by AA-Turner** [comment]
@danieleades if you remove the c/cpp domain changes from this PR (blacklist those modules from the lint) I can merge this.

A

File: sphinx/domains/cpp.py
Comment (jakobandersen) [inline sphinx/domains/cpp.py]: Please change to put ``objectType`` in the exception message.

File: sphinx/domains/cpp.py
Comment (jakobandersen) [inline sphinx/domains/cpp.py]:
```suggestion
```

File: sphinx/domains/cpp.py
Comment (jakobandersen) [inline sphinx/domains/cpp.py]:
```suggestion
```

File: sphinx/domains/cpp.py
Comment (jakobandersen) [inline sphinx/domains/cpp.py]: As above, please change to something similar to ``raise AssertionError(objectType)``.

**Review by jakobandersen — CHANGES_REQUESTED** [review]
As others I think it is a bit icky to change assertions to explicit exceptions, but for the C and C++ domains all the changes are cases of should-not-happen, so it looks basically fine to me.

**Review by jakobandersen — APPROVED** [review]

**Comment by tk0miya** [comment]
Could you check the flake8 warnings, please? I'll merge this if they will be cleared.

File: setup.cfg
Comment (AA-Turner) [inline setup.cfg]:
```suggestion
ignore = E116,E241,E251,E741,W504,I101,B006,B023
```

**Pull request merged**
